verification/spec/connection-handshake/Environment.tla

```diff
@@ -0,0 +1,402 @@
+             ClientIDs      <- { x.clientID : x \in ChainAConnectionEnds }
+
+
+chmB == INSTANCE ConnectionHandshakeModule
+        WITH MaxChainHeight <- MaxHeight,
+             inBuf          <- inBufChainB,
+             outBuf         <- outBufChainB,
+             store          <- storeChainB,
+             ConnectionIDs  <- { x.connectionID : x \in ChainBConnectionEnds },
+             ClientIDs      <- { x.clientID : x \in ChainBConnectionEnds }
+
+
+(***************************************************************************
+ Environment actions.
+ ***************************************************************************)
+
+
+(* Environment initialization.
+    
+    This action kick-stars the ICS3 protocol by assigning an ICS3MsgInit
+    msg to either of the two chains (or both).
+    
+    Initially, the environment is non-malicious. The environment starts
+    acting maliciously once the connection on both chains transitions out
+    of state "UNINIT" (the initials state). It is important to
+    initialize the protocol like this, otherwise the env. can provoke a
+    deadlock. This can happen with the following sequence of actions:
+    
+    1. Environment injects a ICS3MsgInit to chain A with the following
+    correct parameters:
+
+        localEnd |-> [connectionID |-> "connAtoB",
+                      clientID |-> "clientIDChainB"]
+        remoteEnd |-> [connectionID |-> "connBtoA",
+                       clientID |-> "clientIDChainA"]
+                    
+    2. Environment injects, maliciously, a ICS3MsgInit to chain B with
+    the following parameters:
+
+        localEnd |-> [connectionID |-> "connBtoA",
+                      clientID |-> "clientIDChainA"]
+        remoteEnd |-> [connectionID |-> "connBtoA",
+                       clientID |-> "clientIDChainA"]
+                       
+    Notice that the localEnd is correct, so chain B will validate and
+    process this message; the remoteEnd is incorrect, however, but chain
+    B is not able to validate that part of the connection, so it will
+    accept it as it is.
+
+    2. Chain A processes the ICS3MsgInit (action HandleInitMsg) and
+    updates its store.connection with the parameters from step 1 above.
+    At this point, chain A "locks onto" these parameters and will not
+    accept any others. Chain A also produces a ICS3MsgTry message.
+    
+    3. Chain B processes the ICS3MsgInit (action HandleInitMsg) and
+    updates its store.connection with the parameters from step 2 above.
+    Chain B "locks onto" these parameters and will not accept any others.
+    At this step, chain B produces a ICS3MsgTry message with the local
+    parameters from its connection.
+    
+    Both chains will be locked on a different set of connection parameters,
+    and neither chain will accept their corresponding ICS3MsgTry, hence a
+    deadlock. To avoid this problem, we prevent the environment from
```

Comment:
This is really cool scenario; we should be able ideally to capture it with the model checker. Solution to this problem is not constraining environment as we can't do it in practice. We need to design protocol differently otherwise single faulty relayer can prevent us from establishing connections between chains.

Reply:
Opened a separate issue #70 to track discussion on this.